### PR TITLE
Using Domain structure for ContentBlocks module

### PR DESCRIPTION
**Pull request by jeroendesloovere**
## Type

- Enhancement

## Pull request description

> The current ContentBlocks module in Fork CMS uses Doctrine and Symfony forms. That's amazing, but we are going a step further by dividing the logic in a folder called `Domain`. This is the way we are going to build all modules in Fork CMS. If you want to see some examples with the `Domain` folder, you can check out the MediaGalleries and MediaLibrary modules.

### CODE REVIEW

**Comment by jeroendesloovere** [comment]
Note to self: saving a review isn't working

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php [hunk 1]
```diff
@@ -12,26 +12,22 @@
...
         parent::execute();
 
-        $form = $this->createForm(
-            ContentBlockType::class,
-            null,
-            ['theme' => $this->get('fork.settings')->get('Core', 'theme', 'Core')]
-        );
-
-        $form->handleRequest($this->get('request'));
+        /** @var Form $form */
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php]: you don't need this phpdoc since the return type is already the form

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php [hunk 2]
```diff
@@ -12,26 +12,22 @@
...
+use Backend\Modules\ContentBlocks\Domain\ContentBlock\Command\CreateContentBlock;
+use Backend\Modules\ContentBlocks\Domain\ContentBlock\Event\ContentBlockCreated;
+use Backend\Modules\ContentBlocks\Domain\ContentBlock\ContentBlockType;
+use Symfony\Component\Form\Form;
 
 /**
  * This is the add-action, it will display a form to create a new item
  */
-class Add extends BackendBaseActionAdd
+class ContentBlockAdd extends BackendBaseActionAdd
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php]: I think in this case renaming this might be a bit overkill, but if you do it you need to make sure that all the changes needed to the database are added to the upgrade guide

Goes for all the other actions in this case as well.

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php [hunk 3]
```diff
@@ -43,27 +39,55 @@ public function execute(): void
...
+            $parameters
+        );
+    }
+
+    private function getForm(): Form
+    {
+        $form = $this->createForm(
+            ContentBlockType::class,
+            new CreateContentBlock(),
+            ['theme' => $this->get('fork.settings')->get('Core', 'theme', 'Core')]
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php]: This must be `$this->get('fork.settings')->get('Core', 'theme', 'Fork')`

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockDelete.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+use Backend\Modules\ContentBlocks\Domain\ContentBlock\ContentBlock;
+use Backend\Modules\ContentBlocks\Domain\ContentBlock\Event\ContentBlockDeleted;
+use Backend\Modules\ContentBlocks\Domain\ContentBlock\Exception\ContentBlockNotFound;
+
+/**
+ * This is the delete-action, it will delete an item.
+ */
+class ContentBlockDelete extends BackendBaseActionDelete
+{
+    public function execute(): void
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockDelete.php]: if https://github.com/forkcms/forkcms/pull/2090 gets merged before this one the logic here needs to be updated since it won't get merged because you changed everything in one big commit and thereby are breaking the git history

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php [hunk 5]
```diff
@@ -12,26 +12,22 @@
...
-            ContentBlockType::class,
-            null,
-            ['theme' => $this->get('fork.settings')->get('Core', 'theme', 'Core')]
-        );
-
-        $form->handleRequest($this->get('request'));
+        /** @var Form $form */
+        $form = $this->getForm();
 
         if (!$form->isValid()) {
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php]: checking if a form is valid before checking if a form is submitted is deprecated in symfony 3.3 and removed in symfony 4, since we will need to upgrade for fork 5 this will need to be changed here as well

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        parent::execute();
+
+        /** @var ContentBlock $contentBlock */
+        $contentBlock = $this->getContentBlock();
+
+        /** @var Form $form */
+        $form = $this->getForm($contentBlock);
+
+        if (!$form->isValid()) {
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php]: checking if a form is valid before checking if a form is submitted is deprecated in symfony 3.3 and removed in symfony 4, since we will need to upgrade for fork 5 this will need to be changed here as well

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        );
+    }
+
+    private function getContentBlock(): ContentBlock
+    {
+        /** @var ContentBlockRepository $contentBlockRepository */
+        $contentBlockRepository = $this->get('content_blocks.repository.content_block');
+
+        // specific revision?
+        $revisionId = $this->getParameter('revision', 'int');
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php]: $this->getParameter has been removed in fork 5.0.0
use `$this->getRequest()->query->getInt('revision')` instead

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+            try {
+                return $contentBlockRepository->findOneByRevisionIdAndLocale($revisionId, Locale::workingLocale());
+            } catch (ContentBlockNotFound $e) {
+                $this->redirect($this->getBackLink(['error' => 'non-existing']));
+            }
+        }
+
+        try {
+            return $contentBlockRepository->findOneByIdAndLocale($this->getParameter('id', 'int'), Locale::workingLocale());
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php]: $this->getParameter has been removed in fork 5.0.0
use `$this->getRequest()->query->getInt('id')` instead

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        }
+    }
+
+    private function getForm(ContentBlock $contentBlock): Form
+    {
+        $form = $this->createForm(
+            ContentBlockType::class,
+            new UpdateContentBlock($contentBlock),
+            [
+                'theme' => $this->get('fork.settings')->get('Core', 'theme', 'Core'),
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php]: This must be `$this->get('fork.settings')->get('Core', 'theme', 'Fork')`

File: src/Backend/Modules/ContentBlocks/Domain/ContentBlock/Command/CopyContentBlocksToOtherLocaleHandler.php [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        return (array) $this->contentBlockRepository->findBy(
+            [
+                'locale' => $locale,
+                'status' => Status::active()
+            ]
+        );
+    }
+
+    /**
+     * @return int
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Domain/ContentBlock/Command/CopyContentBlocksToOtherLocaleHandler.php]: useless phpdoc

File: src/Backend/Modules/ContentBlocks/Domain/ContentBlock/Command/CopyContentBlocksToOtherLocaleHandler.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        $this->contentBlockRepository = $contentBlockRepository;
+    }
+
+    public function handle(CopyContentBlocksToOtherLocale $copyContentBlocksToOtherLocale): void
+    {
+        /** @var array $contentBlocksToCopy */
+        $contentBlocksToCopy = $this->getContentBlocksToCopy($copyContentBlocksToOtherLocale->fromLocale);
+
+        /** @var int $id */
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Domain/ContentBlock/Command/CopyContentBlocksToOtherLocaleHandler.php]: looks like useless phpdoc to me

File: src/Backend/Modules/ContentBlocks/Domain/ContentBlock/Command/CopyContentBlocksToOtherLocaleHandler.php [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    private $contentBlockRepository;
+
+    public function __construct(ContentBlockRepository $contentBlockRepository)
+    {
+        $this->contentBlockRepository = $contentBlockRepository;
+    }
+
+    public function handle(CopyContentBlocksToOtherLocale $copyContentBlocksToOtherLocale): void
+    {
+        /** @var array $contentBlocksToCopy */
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Domain/ContentBlock/Command/CopyContentBlocksToOtherLocaleHandler.php]: looks like useless phpdoc to me

File: src/Backend/Modules/ContentBlocks/Domain/ContentBlock/ContentBlockDataGrid.php [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        if (BackendAuthentication::isAllowedAction('ContentBlockEdit')) {
+            $editUrl = Model::createURLForAction('ContentBlockEdit', null, null, ['id' => '[id]'], false);
+            $this->setColumnURL('title', $editUrl);
+            $this->addColumn('edit', null, Language::lbl('Edit'), $editUrl, Language::lbl('Edit'));
+        }
+    }
+
+    public static function getHtml(Locale $locale): string
+    {
+        return (string) (new self($locale))->getContent();
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Domain/ContentBlock/ContentBlockDataGrid.php]: cast to string is no longer needed in fork 5 since getContent now always will return a string

File: src/Backend/Modules/ContentBlocks/Domain/ContentBlock/ContentBlockRepository.php [hunk 14]
```diff
@@ -38,20 +38,43 @@ public function getNextIdForLanguage(Locale $locale): int
             ->getSingleScalarResult() + 1;
     }
 
-    public function findOneByIdAndLocale($id, Locale $locale): ?ContentBlock
+    public function findOneByIdAndLocale(int $id = null, Locale $locale): ?ContentBlock
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Domain/ContentBlock/ContentBlockRepository.php]: wrong method signature, if the int needs to be able to receive a null (maybe check the implementation) you need to use `?int` and not `= null` because now you have an optional parameter before a required one

File: src/Backend/Modules/ContentBlocks/Domain/ContentBlock/ContentBlockRepository.php [hunk 15]
```diff
@@ -38,20 +38,43 @@ public function getNextIdForLanguage(Locale $locale): int
...
+
+        if ($contentBlock === null) {
+            throw ContentBlockNotFound::forId($id);
+        }
+
+        return $contentBlock;
     }
 
-    public function findOneByRevisionIdAndLocale($revisionId, Locale $locale): ?ContentBlock
+    public function findOneByRevisionIdAndLocale(int $revisionId = null, Locale $locale): ContentBlock
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Domain/ContentBlock/ContentBlockRepository.php]: likewise

File: src/Backend/Modules/ContentBlocks/Domain/ContentBlock/ContentBlockRevisionDataGrid.php [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+                Language::lbl('UseThisVersion'),
+                $editRevisionUrl,
+                Language::lbl('UseThisVersion')
+            );
+        }
+    }
+
+    public static function getHtml(ContentBlock $contentBlock, Locale $locale): string
+    {
+        return (string) (new self($contentBlock, $locale))->getContent();
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Domain/ContentBlock/ContentBlockRevisionDataGrid.php]: we no longer need to cast to string since `->getContent()` now always will return a string and no longer `null` but `''` when there are no results

File: src/Backend/Modules/ContentBlocks/Domain/ContentBlock/Exception/ContentBlockNotFound.php [hunk 17]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Backend\Modules\ContentBlocks\Domain\ContentBlock\Exception;
+
+use Exception;
+
+class ContentBlockNotFound extends Exception
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Domain/ContentBlock/Exception/ContentBlockNotFound.php]: 👍

File: src/Backend/Modules/ContentBlocks/Layout/Templates/ContentBlockEdit.html.twig [hunk 18]
```diff
@@ -51,7 +51,7 @@
     <div class="col-md-12">
       <div class="btn-toolbar">
         <div class="btn-group pull-left" role="group">
-          {% if showContentBlocksDelete %}
+          {% if showContentBlocksContentBlockDelete %}
```
Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Layout/Templates/ContentBlockEdit.html.twig]: see with the renaming of the actions in this case it only complicates reading things like this

**Comment by carakas** [comment]
I'm going to close this one like we discussed since too much of the git history is broken.
Please look at the feedback and create a new PR with nice descriptive commits and doing one step at a time

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php (hunk 3 above)
Comment (jeroendesloovere) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php]: I guess this has/will be fixed in another PR, so I won't change this. OK?

Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockAdd.php]: That PR has already been made and has been merged, so you should make sure you have the latests version of 5.0.0-dev, since git saw your changes as delete and add instead of move that change would've been lost

File: src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php (hunk 9 above)
Comment (jeroendesloovere) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php]: I guess this has/will be fixed in another PR, so I won't change this. OK?

Comment (carakas) [inline src/Backend/Modules/ContentBlocks/Actions/ContentBlockEdit.php]: same answer here
